In [1]:
import pandas as pd
import numpy as np


In [2]:
df=pd.read_csv('dna_sample/synthetic_dna_dataset.csv')
df.head()

Sample_ID                                           Sequence  GC_Content  \
0  SAMPLE_1  CTTTCGGGATACTTTTGGGATGGTCTTGGTCAAGGGTTTTAGCCCG...        50.0   
1  SAMPLE_2  TTGACCAAATTTGATTGGAAGTGGTAAGCGCGTATTCCTAGCATCA...        45.0   
2  SAMPLE_3  GCGTGAGTTCTAATTTAAAAAGTCGTAACACGTACCCCGGCGTGTA...        51.0   
3  SAMPLE_4  ACTACGCGGACAAGAACCAACAGAACCTGGTTTTCGCAAGGGAGTG...        55.0   
4  SAMPLE_5  TTCAATGCAGATTGAAAGTTACTTTCATCTGCCCTATGGGTCCCTT...        46.0   

   AT_Content  Sequence_Length  Num_A  Num_T  Num_C  Num_G  kmer_3_freq  \
0        50.0              100     22     28     19     31        0.986   
1        55.0              100     27     28     22     23        0.486   
2        49.0              100     26     23     30     21        0.367   
3        45.0              100     28     17     23     32        0.404   
4        54.0              100     24     30     25     21        0.818   

   Mutation_Flag Class_Label Disease_Risk  
0              0    Bacteria         High  
1              1       Virus       Medium  
2              1    Bacteria          Low  
3              0       Human       Medium  
4              0       Human         High

In [3]:
df.drop(labels='Sample_ID',axis=1,inplace=True)
df.head()

Sequence  GC_Content  AT_Content  \
0  CTTTCGGGATACTTTTGGGATGGTCTTGGTCAAGGGTTTTAGCCCG...        50.0        50.0   
1  TTGACCAAATTTGATTGGAAGTGGTAAGCGCGTATTCCTAGCATCA...        45.0        55.0   
2  GCGTGAGTTCTAATTTAAAAAGTCGTAACACGTACCCCGGCGTGTA...        51.0        49.0   
3  ACTACGCGGACAAGAACCAACAGAACCTGGTTTTCGCAAGGGAGTG...        55.0        45.0   
4  TTCAATGCAGATTGAAAGTTACTTTCATCTGCCCTATGGGTCCCTT...        46.0        54.0   

   Sequence_Length  Num_A  Num_T  Num_C  Num_G  kmer_3_freq  Mutation_Flag  \
0              100     22     28     19     31        0.986              0   
1              100     27     28     22     23        0.486              1   
2              100     26     23     30     21        0.367              1   
3              100     28     17     23     32        0.404              0   
4              100     24     30     25     21        0.818              0   

  Class_Label Disease_Risk  
0    Bacteria         High  
1       Virus       Medium  
2    Bacteria          Low  
3       Human       Medium  
4       Human         High

In [4]:
df['Class_Label'].unique()

array(['Bacteria', 'Virus', 'Human', 'Plant'], dtype=object)

In [5]:
df['Class_Label']=df['Class_Label'].map({'Bacteria':1, 'Virus':0, 'Human':3,'Plant':2})
df.head()

Sequence  GC_Content  AT_Content  \
0  CTTTCGGGATACTTTTGGGATGGTCTTGGTCAAGGGTTTTAGCCCG...        50.0        50.0   
1  TTGACCAAATTTGATTGGAAGTGGTAAGCGCGTATTCCTAGCATCA...        45.0        55.0   
2  GCGTGAGTTCTAATTTAAAAAGTCGTAACACGTACCCCGGCGTGTA...        51.0        49.0   
3  ACTACGCGGACAAGAACCAACAGAACCTGGTTTTCGCAAGGGAGTG...        55.0        45.0   
4  TTCAATGCAGATTGAAAGTTACTTTCATCTGCCCTATGGGTCCCTT...        46.0        54.0   

   Sequence_Length  Num_A  Num_T  Num_C  Num_G  kmer_3_freq  Mutation_Flag  \
0              100     22     28     19     31        0.986              0   
1              100     27     28     22     23        0.486              1   
2              100     26     23     30     21        0.367              1   
3              100     28     17     23     32        0.404              0   
4              100     24     30     25     21        0.818              0   

   Class_Label Disease_Risk  
0            1         High  
1            0       Medium  
2            1          Low  
3            3       Medium  
4            3         High

In [6]:
df.isna().value_counts()
#One null/empty cell found

Sequence  GC_Content  AT_Content  Sequence_Length  Num_A  Num_T  Num_C  Num_G  kmer_3_freq  Mutation_Flag  Class_Label  Disease_Risk
False     False       False       False            False  False  False  False  False        False          False        False           3000
Name: count, dtype: int64

In [7]:
#Check if stratified is required
df['Disease_Risk'].value_counts(normalize=True)*100
#evenly distributed thus stratified not required. Still use it to prevent unluck random split

Disease_Risk
High      34.866667
Low       32.600000
Medium    32.533333
Name: proportion, dtype: float64

In [8]:
df['Disease_Risk']=df['Disease_Risk'].map({'High':2,'Medium':1,'Low':0})
df.head()

Sequence  GC_Content  AT_Content  \
0  CTTTCGGGATACTTTTGGGATGGTCTTGGTCAAGGGTTTTAGCCCG...        50.0        50.0   
1  TTGACCAAATTTGATTGGAAGTGGTAAGCGCGTATTCCTAGCATCA...        45.0        55.0   
2  GCGTGAGTTCTAATTTAAAAAGTCGTAACACGTACCCCGGCGTGTA...        51.0        49.0   
3  ACTACGCGGACAAGAACCAACAGAACCTGGTTTTCGCAAGGGAGTG...        55.0        45.0   
4  TTCAATGCAGATTGAAAGTTACTTTCATCTGCCCTATGGGTCCCTT...        46.0        54.0   

   Sequence_Length  Num_A  Num_T  Num_C  Num_G  kmer_3_freq  Mutation_Flag  \
0              100     22     28     19     31        0.986              0   
1              100     27     28     22     23        0.486              1   
2              100     26     23     30     21        0.367              1   
3              100     28     17     23     32        0.404              0   
4              100     24     30     25     21        0.818              0   

   Class_Label  Disease_Risk  
0            1             2  
1            0             1  
2            1             0  
3            3             1  
4            3             2

In [9]:
df.columns

Index(['Sequence', 'GC_Content', 'AT_Content', 'Sequence_Length', 'Num_A',
       'Num_T', 'Num_C', 'Num_G', 'kmer_3_freq', 'Mutation_Flag',
       'Class_Label', 'Disease_Risk'],
      dtype='object')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Sequence         3000 non-null   object 
 1   GC_Content       3000 non-null   float64
 2   AT_Content       3000 non-null   float64
 3   Sequence_Length  3000 non-null   int64  
 4   Num_A            3000 non-null   int64  
 5   Num_T            3000 non-null   int64  
 6   Num_C            3000 non-null   int64  
 7   Num_G            3000 non-null   int64  
 8   kmer_3_freq      3000 non-null   float64
 9   Mutation_Flag    3000 non-null   int64  
 10  Class_Label      3000 non-null   int64  
 11  Disease_Risk     3000 non-null   int64  
dtypes: float64(3), int64(8), object(1)
memory usage: 281.4+ KB


In [11]:
df.drop(labels='Sequence', axis=1,inplace=True)

### Predicting Varible is 'Disease Risk'.

- The prediction have 3 category, thus binary classifications arent good , exclude logistic regression
- svm is great for high dimension and textual data however most of the data is numeric 10 dim, SVM have potential but not top priority
- Random Forest Classification is good for complex data and is a combination of multiple decision tree, so it looks good
- knn good for low dimension and low complex data, not good in this case
- decision tree may also perform good , but might not as good as random forest tree, so its on low priority
- Naive bayes works on very high dim where dimension>no. of rows, In this case the DNA might be co-related and naive bayes treats each attribute indivisually for prediction, thus might not be the best, so low priority

### Cross validation will be done on models
1) Random Forest
2) Decision Tree
3) XGBoost
4) ADABoost
5) Gradient Boost

Splitting of dataset needs to be considered.

    Since we are doing classification we will use Stratified split, and also need to use cross-validation
    Thus we will use  'Stratified K-Fold' Splitting

*Note:* We are using all trees since this dataset might be best with it , also the encoding is done using map, which is good for tree based model

In [20]:
from sklearn.model_selection import StratifiedKFold,cross_val_score,RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier,AdaBoostClassifier,GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier
import xgboost as xgb

In [13]:
x=df.drop(columns=['Disease_Risk'])
y=df['Disease_Risk']

In [16]:
#Define Models
models={
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random forset": RandomForestClassifier(random_state=42),
    "XGBoost" : xgb.XGBClassifier(eval_metric='mlogloss', random_state=42),
    "AdaBoost": AdaBoostClassifier(random_state=42),
    "Gradient Boost":GradientBoostingClassifier(random_state=42)
}

#For now train all model in default hyper paraneter to see how it looks

In [17]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

for name,model in models.items():
    score=cross_val_score(model,x,y,cv=skf,scoring='accuracy') #Other scoring attributes can also be used
    print(f'{name} :Mean accuracy={score.mean():4f},std={score.std():.4f}')

Decision Tree :Mean accuracy=0.339000,std=0.0062
Random forset :Mean accuracy=0.344000,std=0.0247
XGBoost :Mean accuracy=0.340000,std=0.0228
AdaBoost :Mean accuracy=0.342333,std=0.0155
Gradient Boost :Mean accuracy=0.347667,std=0.0174


In [19]:
for name,model in models.items():
    score=cross_val_score(model,x,y,cv=skf,scoring='f1_macro') #Other scoring attributes can also be used
    print(f'{name} :Mean f1_macro={score.mean():4f},std={score.std():.4f}')

Decision Tree :Mean f1_macro=0.339010,std=0.0063
Random forset :Mean f1_macro=0.343443,std=0.0247
XGBoost :Mean f1_macro=0.339395,std=0.0230
AdaBoost :Mean f1_macro=0.292403,std=0.0220
Gradient Boost :Mean f1_macro=0.342483,std=0.0172


# Selecting Model
- AdaBoost had good accuracy but failed in f1_macro(clearly worse)
- Decision Tree decent but not good
- Gradient Boost, very less std->stable, with good f1macro and accuracy
- XGBoost better model than Gradient in default 
- Random Forest clearly the best model over here in terms of accuracy and f1_score


# We will be choosing 1 model for simplification, practically 2-3 models will be taken and hyper-tunned.

*Note: Having a  accuracy(34%) shows that there might not be a strong co-relation between independent and dependent variables. Thus hyperparameter tunning wont drastically increase the accuracy.*

In [23]:
#Use "GridSearch cv" or "RandomizeSearch cv" for hyper_parameter tunning. Practically Randomize is better since it wont brute force all combination but might miss the best combo
rf=RandomForestClassifier(random_state=42)

parameter_grid={
    'n_estimators': [200,400,600],
    'max_depth':[5,10,15,None],
    'min_samples_split':[2,5,10],
    'min_samples_leaf':[1,2,4],
    'max_features':['sqrt','log2']
}

search= RandomizedSearchCV(rf,parameter_grid,n_iter=30,cv=5,scoring='f1_macro',n_jobs=-1)

| Argument              | Meaning                                                                           |
| --------------------- | --------------------------------------------------------------------------------- |
| `rf`                  | The **estimator** (RandomForestClassifier) to tune                                |
| `param_distributions` | The **hyperparameter dictionary** you defined — tells it what values to pick from |
| `n_iter=30`           | RandomizedSearchCV will try **30 random combinations** from the dictionary        |
| `cv=5`                | 5-fold cross-validation for each combination                                      |
| `scoring="f1_macro"`  | Use F1-macro to evaluate performance                                              |
| `n_jobs=-1`           | Use all CPU cores for faster computation                                          |
| `random_state=42`     | Makes the random sampling reproducible                                            |


In [24]:
search.fit(x,y)
print(search.best_score_)
print(search.best_params_)

0.33234271214118705
{'n_estimators': 400, 'min_samples_split': 5, 'min_samples_leaf': 2, 'max_features': 'sqrt', 'max_depth': None}


# We hit the dead-end with tree based model we cant compute any further
*Note: We have dropped sequential dna sequence so that might be the reason for low performance, however dna sequence is better to be trained with deep-learning, thus excluded in this proj.*

Best accuracy(34.4%) and f1macro score(34.3%) was obtained in random forest with default parameter 